### ISSUE TITLE
BUG: App crashes when adding notes after deleting a point

### ISSUE THREAD

**Issue by tomarney** [BritishGeologicalSurvey/tactool#9]
The app keeps crashing when adding notes to points after deleting a point.

Steps to reproduce:

1. run `main.py --dev` (or launch the pre-built app and load any image)
2. click anywhere in the image (sets a RefMark)
3. select the point in the table and add a note to it
4. lay a second point, **then right-click it to remove it**
5. add another point and add a note to it
6. add a third point and add a note

As soon as the notes cell in the table loses focus, the app will crash with the following error:

```python
(tactool-windows) PS C:\repos\BritishGeologicalSurvey\tactool> c:\Users\tomar\miniconda3\envs\tactool-windows\python.exe tactool/main.py --dev
2023-08-21 13:01:14,816 tactool Initialising TACtool application
2023-08-21 13:01:23,091 tactool Created Analysis Point: AnalysisPoint(id=1, label='RefMark', x=109, y=88, diameter=10, scale=1.0, colour='#ffff00', sample_name='', mount_name='', material='', notes='', _outer_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B028C0>, _inner_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B02830>, _label_text_item=<PyQt5.QtWidgets.QGraphicsTextItem object at 0x00000173A6B02950>)
2023-08-21 13:01:24,146 tactool Selected Analysis Point: AnalysisPoint(id=1, label='RefMark', x=109, y=88, diameter=10, scale=1.0, colour='#ffff00', sample_name='', mount_name='', material='', notes='', _outer_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B028C0>, _inner_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B02830>, _label_text_item=<PyQt5.QtWidgets.QGraphicsTextItem object at 0x00000173A6B02950>)
2023-08-21 13:01:27,406 tactool Deleted Analysis Point: 1
2023-08-21 13:01:27,407 tactool Created Analysis Point: AnalysisPoint(id=1, label='RefMark', x=109, y=88, diameter=10, scale=1.0, colour='#ffff00', sample_name='', mount_name='', material='', notes='my note', _outer_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B029E0>, _inner_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B027A0>, _label_text_item=<PyQt5.QtWidgets.QGraphicsTextItem object at 0x00000173A6B02A70>)
2023-08-21 13:01:29,038 tactool Created Analysis Point: AnalysisPoint(id=2, label='RefMark', x=160, y=161, diameter=10, scale=1.0, colour='#ffff00', sample_name='', mount_name='', material='', notes='', _outer_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B028C0>, _inner_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B02830>, _label_text_item=<PyQt5.QtWidgets.QGraphicsTextItem object at 0x00000173A6B02B00>)
2023-08-21 13:01:30,859 tactool Deleted Analysis Point: 2
2023-08-21 13:01:31,519 tactool Created Analysis Point: AnalysisPoint(id=3, label='RefMark', x=160, y=158, diameter=10, scale=1.0, colour='#ffff00', sample_name='', mount_name='', material='', notes='', _outer_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B02C20>, _inner_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B028C0>, _label_text_item=<PyQt5.QtWidgets.QGraphicsTextItem object at 0x00000173A6B02B00>)
2023-08-21 13:01:32,557 tactool Selected Analysis Point: AnalysisPoint(id=3, label='RefMark', x=160, y=158, diameter=10, scale=1.0, colour='#ffff00', sample_name='', mount_name='', material='', notes='', _outer_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B02C20>, _inner_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B028C0>, _label_text_item=<PyQt5.QtWidgets.QGraphicsTextItem object at 0x00000173A6B02B00>)
2023-08-21 13:01:43,975 tactool Deleted Analysis Point: 1
2023-08-21 13:01:43,976 tactool Deleted Analysis Point: 3
2023-08-21 13:01:43,978 tactool Created Analysis Point: AnalysisPoint(id=1, label='RefMark', x=109, y=88, diameter=10, scale=1.0, colour='#ffff00', sample_name='', mount_name='', material='', notes='my note', _outer_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B02830>, _inner_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B02B90>, _label_text_item=<PyQt5.QtWidgets.QGraphicsTextItem object at 0x00000173A6B02D40>)
2023-08-21 13:01:43,980 tactool Created Analysis Point: AnalysisPoint(id=3, label='RefMark', x=160, y=158, diameter=10, scale=1.0, colour='#ffff00', sample_name='', mount_name='', material='', notes='my 2nd note', _outer_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B02DD0>, _inner_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B02E60>, _label_text_item=<PyQt5.QtWidgets.QGraphicsTextItem object at 0x00000173A6B02EF0>)
2023-08-21 13:01:43,985 tactool Created Analysis Point: AnalysisPoint(id=3, label='RefMark', x=197, y=218, diameter=10, scale=1.0, colour='#ffff00', sample_name='', mount_name='', material='', notes='', _outer_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B028C0>, _inner_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B02B00>, _label_text_item=<PyQt5.QtWidgets.QGraphicsTextItem object at 0x00000173A6B02C20>)
2023-08-21 13:01:55,977 tactool Selected Analysis Point: AnalysisPoint(id=3, label='RefMark', x=197, y=218, diameter=10, scale=1.0, colour='#ffff00', sample_name='', mount_name='', material='', notes='', _outer_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B028C0>, _inner_ellipse=<PyQt5.QtWidgets.QGraphicsEllipseItem object at 0x00000173A6B02B00>, _label_text_item=<PyQt5.QtWidgets.QGraphicsTextItem object at 0x00000173A6B02C20>)
2023-08-21 13:02:04,774 tactool Deleted Analysis Point: 1
Traceback (most recent call last):
  File "C:\repos\BritishGeologicalSurvey\tactool\tactool\window.py", line 668, in update_analysis_points
    self.reload_analysis_points()
  File "C:\repos\BritishGeologicalSurvey\tactool\tactool\window.py", line 607, in reload_analysis_points
    self.clear_analysis_points()
  File "C:\repos\BritishGeologicalSurvey\tactool\tactool\window.py", line 658, in clear_analysis_points
    self.remove_analysis_point(apid=point.id)
  File "C:\repos\BritishGeologicalSurvey\tactool\tactool\window.py", line 591, in remove_analysis_point
    deletion_result = self.graphics_scene.remove_analysis_point(x=x, y=y, apid=apid)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\repos\BritishGeologicalSurvey\tactool\tactool\graphics_scene.py", line 153, in remove_analysis_point
    self.table_model.remove_point(analysis_point.id)
  File "C:\repos\BritishGeologicalSurvey\tactool\tactool\table_model.py", line 228, in remove_point
    self._data.pop(index)
IndexError: pop index out of range
```

Here's a screenshot of the app state immediately before the notes cell loses focus:

![screenshot](https://github.com/BritishGeologicalSurvey/tactool/assets/43818254/8dae3d2f-ef52-4613-ab4c-c2cd5efa7c2c)

The second and third points both have the same ID (`3`).

I have reproduced this with different notes, using spots instead of RefMarks, having set the scale, etc. It seems the critical step is adding a note to the second and third points in this example.

**Comment by tomarney** [comment]
As a workaround, if you delete a point, click "Reset IDs" button immediately and adding notes should work as expected.

**Comment by leorudczenko** [comment]
Hi Tom! Thanks for reporting this bug with all the information, it is very helpful. I was able to reproduce the same issue on the `main` branch.

We are currently working on adding a re-coordination feature to TACtool, you can view the current progress on the branch relating to issue #4. Some of the internal logic has been improved, and currently this bug seems to have gone with the logic changes on that branch. To prevent a similar bug reoccurring, I have added some extra test cases for the functions which caused the issue.

In summary, this bug will fixed in the next release of TACtool!
